cpu/nrf51822/periph/flash.c

```diff
@@ -0,0 +1,203 @@
+ *       NRF_FICR->CLENR0 + PPFC
+ */
+static uint8_t flash_check_address(void *address)
+{
+    if (((size_t)address < 0) || ((size_t)address >= _nrf51_flash_size)) {
+        return FLASH_ERROR_ADDR_RANGE;
+    }
+
+    return FLASH_ERROR_SUCCESS;
+}
+
+
+
+/************************************************************************/
+/* periph/flash.h *******************************************************/
+/************************************************************************/
+
+uint8_t flash_init(void)
+{
+    DEBUG("flash initialized\n");
+    return FLASH_ERROR_SUCCESS;
+}
+
+flash_page_number_t flash_get_page_number(void *address, flash_page_size_t *page_offset)
+{
+    flash_page_number_t page_number = (flash_page_number_t)(
+                                          (size_t)address / FLASH_PAGE_SIZE);
+
+    if (page_offset != NULL) {
+        *page_offset = (flash_page_number_t)(
+                           (size_t)address % FLASH_PAGE_SIZE);
+    }
+
+    return page_number;
+}
+
+void *flash_get_address(flash_page_number_t page)
+{
+    return ((void *)(page * FLASH_PAGE_SIZE));
+}
+
+uint8_t flash_memcpy(void *dest, const void *src, size_t n)
+{
+    /* Check alignment */
+    flash_page_size_t page_offset;
+    flash_get_page_number(dest, &page_offset);
+
+    if ((n % FLASH_WRITE_ALIGN != 0) || (page_offset % FLASH_WRITE_ALIGN != 0)) {
+        DEBUG("Unaligned access n=%d offset=%d\n", n, page_offset);
+        return FLASH_ERROR_ALIGNMENT;
+    }
+
+    /* write data to flash */
+    uint8_t ret = flash_memcpy_fast(dest, src, n);
+
+    /* verify written data if write was successfull */
+    if (ret == FLASH_ERROR_SUCCESS) {
+        while (n > 0) {
+            if (*((flash_data_word_t *) dest) != *((flash_data_word_t *) src)) {
+                return FLASH_ERROR_VERIFY;
+            }
+
+            dest += FLASH_WRITE_ALIGN;
+            src  += FLASH_WRITE_ALIGN;
+            n    -= FLASH_WRITE_ALIGN;
+        }
+    }
+
+    return (ret);
+}
+
+/* TODO: Fix it. Jumps into Hard Fault when src is Pointer to char* */
+uint8_t flash_memcpy_fast(void *dest, const void *src, size_t n)
+{
+    /* Check memory range */
+    uint8_t ret = flash_check_address((uint8_t *) dest);
+
+    if (ret > FLASH_ERROR_SUCCESS) {
+        DEBUG("attempted to write below first address\n");
+        return ret;
+    }
+
+    ret = flash_check_address((((uint8_t *) dest) + n));
+
+    if (ret > FLASH_ERROR_SUCCESS) {
+        DEBUG("attemted to write beyond last address\n");
+        return ret;
+    }
+
+    flash_data_word_t *mydst = (flash_data_word_t *)dest;
+    flash_data_word_t *mysrc = (flash_data_word_t *)src;
+
+    /* Enable writing to flash */
+    NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Wen << NVMC_CONFIG_WEN_Pos;
+
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+
+    /* Write data to flash */
+    n = n / FLASH_WRITE_ALIGN;
+
+    while (n--) {
+        *mydst = *mysrc;
```

Comment:
Here it dies with Hard Fault when src points char * in test/periph_flash/main.c. When src is another flash page it works. Instead of reading src it works also when i write an flash_data_word_t variable, but it dies when i read from src into this variable and write it to dst.

I have searched for many hours where the problem is. I have no idea at the moment. Who can help?